### PR TITLE
Specialisation of mul! for left multiplication with scalars

### LINKED ISSUE #181: Doc Instructions for Transpose Insufficient for Full Functionality
I've run into an issue where the instructions on creating transposes for custom linear maps are insufficient for the custom map to fully work. A MWE using the documented example of `MyFillMap`:

<details>

```julia
using LinearMaps, LinearAlgebra

struct MyFillMap{T} <: LinearMaps.LinearMap{T}
    λ::T
    size::Dims{2}
    function MyFillMap(λ::T, dims::Dims{2}) where {T}
        all(≥(0), dims) || throw(ArgumentError("dims of MyFillMap must be non-negative"))
        promote_type(T, typeof(λ)) == T || throw(InexactError())
        return new{T}(λ, dims)
    end
end

Base.size(A::MyFillMap) = A.size

function LinearAlgebra.mul!(y::AbstractVecOrMat, A::MyFillMap, x::AbstractVector)
    LinearMaps.check_dim_mul(y, A, x)
    return fill!(y, iszero(A.λ) ? zero(eltype(y)) : A.λ*sum(x))
end

function LinearAlgebra.mul!(
    y::AbstractVecOrMat,
    transA::LinearMaps.TransposeMap{<:Any,<:MyFillMap},
    x::AbstractVector
)
    LinearMaps.check_dim_mul(y, transA, x)
    λ = transA.lmap.λ
    return fill!(y, iszero(λ) ? zero(eltype(y)) : transpose(λ)*sum(x))
end

```

</details>

Then,

```julia
julia> let M = MyFillMap(1, (5, 5))' * LinearMaps.UniformScalingMap(1, (5, 5))
         M * zeros(5)
       end
ERROR: transpose not implemented for 5×5 MyFillMap{Int64}
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] _unsafe_mul!(y::Vector{Float64}, A::LinearMaps.TransposeMap{Int64, MyFillMap{Int64}}, x::Vector{Float64})
   @ LinearMaps ~/.julia/packages/LinearMaps/NF1pj/src/transpose.jl:54
 [3] _compositemul!(y::Vector{Float64}, A::LinearMaps.CompositeMap{Int64, Tuple{LinearMaps.UniformScalingMap{Int64}, LinearMaps.TransposeMap{Int64, MyFillMap{Int64}}}}, x::Vector{Float64}, source::Vector{Float64}, dest::Nothing)
   @ LinearMaps ~/.julia/packages/LinearMaps/NF1pj/src/composition.jl:186
 [4] _compositemul!(y::Vector{Float64}, A::LinearMaps.CompositeMap{Int64, Tuple{LinearMaps.UniformScalingMap{Int64}, LinearMaps.TransposeMap{Int64, MyFillMap{Int64}}}}, x::Vector{Float64})
   @ LinearMaps ~/.julia/packages/LinearMaps/NF1pj/src/composition.jl:185
 [5] _unsafe_mul!(y::Vector{Float64}, A::LinearMaps.CompositeMap{Int64, Tuple{LinearMaps.UniformScalingMap{Int64}, LinearMaps.TransposeMap{Int64, MyFillMap{Int64}}}}, x::Vector{Float64})
   @ LinearMaps ~/.julia/packages/LinearMaps/NF1pj/src/composition.jl:168
 [6] mul!(y::Vector{Float64}, A::LinearMaps.CompositeMap{Int64, Tuple{LinearMaps.UniformScalingMap{Int64}, LinearMaps.TransposeMap{Int64, MyFillMap{Int64}}}}, x::Vector{Float64})
   @ LinearMaps ~/.julia/packages/LinearMaps/NF1pj/src/LinearMaps.jl:163
 [7] *(A::LinearMaps.CompositeMap{Int64, Tuple{LinearMaps.UniformScalingMap{Int64}, LinearMaps.TransposeMap{Int64, MyFillMap{Int64}}}}, x::Vector{Float64})
   @ LinearMaps ~/.julia/packages/LinearMaps/NF1pj/src/LinearMaps.jl:131
 [8] top-level scope
   @ REPL[5]:2
```

Adding this:

```julia
function LinearMaps._unsafe_mul!(
    y::AbstractVecOrMat,
    transA::LinearMaps.TransposeMap{<:Any,<:MyFillMap},
    x::AbstractVector
)
    λ = transA.lmap.λ
    fill!(y, iszero(λ) ? zero(eltype(y)) : transpose(λ)*sum(x))
end
```

Makes the example work.

### PR DESCRIPTION
**Pull request by krcools**
In `Base.LinearAlgebra`, `mul!` has specialisations for when `s` in `mul!(C,s,B,alpha,beta)` is a scalar. This is needed to make mul! essentially act like inplace addition (add! does not seem to exist and `add!(C,B)` is equivalent to `mul!(C,1,B,1,1)`).

I suggest to echo the treatment of this special but important case also in LinearMaps. As a motivation, what I want is the conversion of a LinearCombination to a Matrix, with minimal overhead:

```
function convert_to_dense(A::LinearMaps.LinearCombination)
    M = zeros(eltype(A), axes(A))
    for map in A.maps
        mul!(M, 1, map, 1, 1)
    end
    return M
end
```

as opposed to

```
function convert_to_dense(A::LinearMaps.LinearCombination)
    M = zeros(eltype(A), axes(A))
    for map in A.maps
        M .+= Matrix(map)
    end
    return M
end
```

The latter does a lot of spurious copying and allocating, and does not allow for clever specialisations in case `map` is a `WrappedMap` around a sparse matrix.

Closes #181.

### CODE REVIEW

**Comment by dkarrasch** [comment]
Also, since this is a new feature, feel free to bump the minor version right away, and leave a comment in the `history.md` file to briefly mention the feature.

**Comment by krcools** [comment]
I am getting some interesting results. The method at

https://github.com/JuliaLinearAlgebra/LinearMaps.jl/blob/f8eaf9846f865f74ddb5dda831081f6a6e8db734/src/conversion.jl#L76

Is only called when all coefficients in the linear combination are unity and when the linear combination is entered *by hand*, i.e. when the storage type is `Tuple`, not `Vector`. This implementation beats my code path by about 12%. My guess is that it called `BLAS.axpy` under the hood...

However, the conditions for entering this code path are very brittle, and in all other cases the fallback implementation is 300 slower. In terms of memory my method is on par.

```
using LinearAlgebra
using LinearMaps
using BenchmarkTools

n = 1024
A = [rand(n,n) for i in 1:10]
L = [LinearMap(a) for a in A]
L2 = [2*LinearMap(a) for a in A]

LC1 = sum(L)
LC2 = sum(L2)

LC3 = L[1] + L[2] + L[3] + L[4] + L[5] + L[6] + L[7] + L[8] + L[9] + L[10]
LC4 = 2L[1] + L[2] + L[3] + L[4] + L[5] + L[6] + L[7] + L[8] + L[9] + L[10]

function convert_to_dense(A::LinearMaps.LinearCombination)
    M = zeros(eltype(A), axes(A))
    for map in A.maps
        mul!(M, 1, map, 1, 1)
    end
    return M
end

@which Matrix{Float64}(LC1)
@which Matrix{Float64}(LC2)
@which Matrix{Float64}(LC3)
@which Matrix{Float64}(LC4)

@btime B1 = Matrix{Float64}(LC1);
@btime B2 = Matrix{Float64}(LC2);
@btime B2 = Matrix{Float64}(LC3);
@btime B3 = Matrix{Float64}(LC4);

@btime C1 = convert_to_dense(LC1);
@btime C2 = convert_to_dense(LC2);
@btime C3 = convert_to_dense(LC3);
@btime C4 = convert_to_dense(LC4);
```

```
  2.618 s (10243 allocations: 8.32 MiB)
  2.644 s (11267 allocations: 16.45 MiB)
  7.000 ms (30 allocations: 16.00 MiB)
  2.614 s (1027 allocations: 16.13 MiB)
  8.642 ms (32 allocations: 8.00 MiB)
  8.547 ms (32 allocations: 8.00 MiB)
  8.608 ms (32 allocations: 8.00 MiB)
  8.467 ms (51 allocations: 8.00 MiB)
```

File: src/scaledmap.jl [hunk 1]
```diff
@@ -64,3 +64,6 @@ for (In, Out) in ((AbstractVector, AbstractVecOrMat),
         end
     end
 end
+
+mul!(Y::AbstractMatrix, c::Number, X::ScaledMap, a::Number, b::Number) =
+    mul!(Y, c, X.lmap, a*X.λ, b)
```
Comment (dkarrasch) [inline src/scaledmap.jl:69]: This should perhaps be moved to the `(In, Out)`-loop above and adjusted to the situation there. Rearranging factors requires the types to belong to some commutative class like `RealOrComplex`.

**Comment by dkarrasch** [comment]
This is getting very close to what I imagined. I think we should make the most generic `Matrix(::LinearMap)` fallback use `generic_mapnum_mul!` and use exactly the previous code in there. In particular, don't overwrite in each loop iteration the whole vector `xi` with zeros, but only overwrite the current entry. Then we have a fully operational fallback even for the wildest custom map types. Given that, you may then be able to simplify (or even improve) the code in a few places. For instance, statements like `Y .+= Matrix(L)` "hurt my eyes", because we allocate the entire matrix on the right, and then copy it to `Y`. I would guess/hope that we can avoid that if make it write into `Y` columnwise directly via the `generic...` code?

**Comment by krcools** [comment]
Fully agree. Came to the same conclusion as I progressed. I will implement these suggestions in the next couple of days.

**Comment by krcools** [comment]
There is actually one thing I can't figure out:

If a user defines a custom LinearMap CustomMap in an external package, the docs say that implementing mul! suffices. However, assume that this custom map is then scaled, giving rise to `ScaledMap{T, CustomMap}` and mul! is called on this scaled map, the following stack unfolds:

```
mul!(y, S::ScaledMap, x)
unsafe_mul!(y, S::ScaledMap, x)
unsafe_mul!(y, C::LinearMap, x)
__generic_mapvec_mul!(y, C::CustomMap, x)
```

In other words, when the custom map appears at a lower level of a more complicated LinearMap, the user supplied `mul!(y, C::CustomMap, x)` is not called. I encounter this issue also in the context of mul! by scalars, but it is not unique to that situation.

Am I missing something here?

Currently I am working around this by not only supplying mul! but also _unsafe_mul! in LiftedMaps.jl.

**Comment by dkarrasch** [comment]
Can you provide a minimal example? Because of

https://github.com/JuliaLinearAlgebra/LinearMaps.jl/blob/aa54eb38bf1863786b59417622a22f81ca8a9253/src/LinearMaps.jl#L246-L247

a "missing" `_unsafe_mul!` should be redirected back to `mul!`. Even if not, then in `_generic_mapvec_mul!` the 3-arg `mul!` gets called. OTOH, `_generic_mapvec_mul!` is called only in the 5-arg case here:

https://github.com/JuliaLinearAlgebra/LinearMaps.jl/blob/aa54eb38bf1863786b59417622a22f81ca8a9253/src/LinearMaps.jl#L248-L250

so, unless I am missing some path, it should really end up with the user-defined `mul!`.

**Comment by krcools** [comment]
I see... I didn't bother drilling down deeper after encountering `__generic_mapvec_mul!`.

The problem still exist for mapmat and mapnum I think. For example, `__generic_mapmat_mul!` is implemented in terms of `mul!(x, L, y::AbstractVector)`, skipping user supplied specialisations of this level BLAS 3 type of operation.

MWE:

```
using LinearMaps

struct CustomMap{T} <: LinearMap{T}
    A::Matrix{T}
end

LinearMaps.MulStyle(C::CustomMap) = LinearMaps.FiveArg()
LinearMaps.size(C::CustomMap) = size(C.A)
LinearMaps.mul!(y::AbstractVector, C::CustomMap, x::AbstractVector, a::Number, b::Number) = LinearMaps.mul!(y, C.A, x, a, b)
function LinearMaps.mul!(Y::AbstractMatrix, C::CustomMap, X::AbstractMatrix, a::Number, b::Number)
    println("User supplied BLAS-3")
    LinearMaps.mul!(Y, C.A, X, a, b)
end

A = rand(3,3)
C = CustomMap(A)
S = 2C

X = rand(3,4)
Y = zeros(3,4)

LinearMaps.mul!(Y,C,X,1,1);
LinearMaps.mul!(Y,S,X,1,1);
```

The same is true for matnum, which is written also written in terms of `mul!(x, L, y::AbstractVector)`, skipping `mul!(x, L, s::Number)`.

It seems to me that `__generic` means something different for mapvec on one hand and for mapmat and matnum on the other. For the former, it exploits special values of alpha and beta, for the latter it is a true fallback that breaks the operation down in primitives that always need to be supplied.

**Comment by dkarrasch** [comment]
Awesome, thanks for finding and sharing that. There are two things that bite each other here. The `_unsafe_mul!` redirection was meant to save size checks for all "submaps", when you, due to checks at construction, should already know that everything is consistent. If we replaced all those by `mul!`, then all this effort is in vain. Could we then solve the issue by asking users to define `_unsafe_mul!` for custom map types? Like tell them to make important checks at construction, then leave the size compatibility to LinearMaps.jl, and then only define the "multiplication kernel" `_unsafe_mul!`? As you say, the good thing is that this affects currently only the mapmat case (and mapnum in the future).

**Comment by dkarrasch** [comment]
> Currently I am working around this by not only supplying mul! but also _unsafe_mul! in LiftedMaps.jl.

Yea, perhaps only `unsafe_mul!` should be required, and all the "high-level" stuff like allocating the output in `A*x` or checking sizes in `mul!` is left to LinearMaps.jl?

**Comment by krcools** [comment]
Yes, that seems to be the way out. Just a shame `_unsafe_mul!` is not a particularly inviting name to import and extend...

File: src/uniformscalingmap.jl [hunk 2]
```diff
@@ -70,6 +70,14 @@ for (In, Out) in ((AbstractVector, AbstractVecOrMat), (AbstractMatrix, AbstractM
     end
 end
 
+function _unsafe_mul!(M::AbstractMatrix, L::UniformScalingMap, s::Number, a::Number, b::Number)
+    rmul!(M,b)
+    for (i,j) in zip(axes(L)...)
+        M[i,j] += L.λ * s * a
```
Comment (dkarrasch) [inline src/uniformscalingmap.jl:76]: Probably this gets hoisted out of the loop, but we could as well make sure manually that the rhs gets computed only once.

**Review by dkarrasch — COMMENTED** [review]
I left a few remarks and suggestions. This looks pretty good, and we should get it done and out soon.

File: src/conversion.jl [hunk 3]
```diff
@@ -1,21 +1,22 @@
...
-    v = fill(zero(T), N)
-    @inbounds for i in 1:N
-        v[i] = one(T)
-        # need mul!, e.g., for TransposeMap{<:CustomMap}
-        mul!(view(mat, :, i), A, v)
-        v[i] = zero(T)
-    end
-    return mat
+    mat = zeros(T, size(A))
+    mul!(mat, A, one(T), one(T), one(T))
```
[the remark this answers is not in the document]
Comment (krcools) [reply src/conversion.jl:4]: I think the options are

```
similar(AbstractArray{T}, axes) # followed by beta=false
similar(Array{T}, axes) # followed by beta=true
```

For blockaxes the former gives a BlockArray{T}, whereas the latter gives a PseudoBlockArray{T}. Not sure what the best choice is. Maybe we best provide full support of similar; this gives full control over both the storage type and the axes type.

File: src/conversion.jl [hunk 4]
```diff
@@ -48,13 +49,10 @@ SparseArrays.SparseMatrixCSC(A::LinearMap) = sparse(A)
 # special cases
 
 # ScaledMap
-Base.Matrix{T}(A::ScaledMap{<:Any, <:Any, <:VecOrMatMap}) where {T} =
-    convert(Matrix{T}, A.λ * A.lmap.lmap)
```
[the remark this answers is not in the document]
Comment (krcools) [reply src/conversion.jl:52]: They are now dealt with by Matrix{T} on line 2. The subtype specific code I removed here can be found in the relevant implementations of _unsafe_mul, so no loss of efficiency is incurred.

**Comment by dkarrasch** [comment]
For me, all your benchmarks from https://github.com/JuliaLinearAlgebra/LinearMaps.jl/pull/173#issuecomment-1084749052 are now on the order of milliseconds, with very small overhead for the linear combination of the doubled maps. I used this code for benchmarking:

```julia
using LinearAlgebra, LinearMaps, BenchmarkTools

n = 1024
A = [rand(n,n) for i in 1:10]

LC1 = sum((LinearMap(a) for a in A)) # creates a LinearMapTuple
LC2 = sum((2*LinearMap(a) for a in A)) # creates a LinearMapTuple
LC3 = sum([LinearMap(a) for a in A]) # creates a LinearMapVector
LC4 = sum([2*LinearMap(a) for a in A]) # creates a LinearMapVector

@btime B1 = Matrix{Float64}(LC1);
@btime B2 = Matrix{Float64}(LC2);
@btime B2 = Matrix{Float64}(LC3);
@btime B3 = Matrix{Float64}(LC4);

  12.059 ms (29 allocations: 8.00 MiB)
  12.252 ms (29 allocations: 8.00 MiB)
  12.147 ms (39 allocations: 8.00 MiB)
  12.884 ms (39 allocations: 8.00 MiB)
```

I agree that the slight regression for the one special case is acceptable, given that this "recurses" into the `_unsafe_mul!(::Matrix, ::LinearMap, ::Number)`  hierarchy and makes use of possible optimizations at every step of the map hierarchy, hence the very little overhead for the linear combination of scaled matrices. As you say, this will further auto-optimize for sparse matrices etc. I also like that with this approach, we first allocate the target, and then map into it, rather than possibly creating intermediate arrays and then stack them together as needed. Local tests suggested that this is also advantageous for `BlockMaps`. And finally, we missed to provide some specialized `Matrix` constructor for our new `EmbeddedMap`s, so relative to the naive implementation, this will speed up things as we avoid computing all-zero columns.

**Pull request merged**